### DISCUSSION TITLE
Cannot mock vue-router

### DISCUSSION THREAD

**Discussion by ryinner** [vitest-dev/vitest#2243] [Q&A] [answered]
I'm testing page with dynamic routing.

```JavaScript
const route = useRoute();
// code
watch(
    () => route.params.id,
    () => getData(),
);
```

I tried mock `vue-router`, but component return undefined.

```JavaScript
vi.mock('vue-router', () => {
    return {
        useRoute: vi.fn(),
    }
});
describe("Test", () => {
    test("Test", () => {
        useRoute.mockImplementationOnce(() => {
            return {
                params: {
                    id: 57759
                }
            }
        });
        const {getByText} = render(Component);
        getByText("test")
   })
})
```

Test result:
`TypeError: Cannot read properties of undefined (reading 'params')`.

If i mock another library, for example axios, that will be work.

```JavaScript
// Working
vi.mock('@/Config/Axios', () => ({
    default: {
        get: vi.fn().mockImplementationOnce(() => {
            return {};
        })
    }
}))
```

I use: 
- testing-library/vue 6.6.1;
- vue 3.2.41
- vue-router 4.1.5
- vitejs/plugin-vue 3.2.0
- vitest 0.24.3

**Comment by sheremet-va** [comment, marked as answer]
Try updating to Vitest 0.24.4. Also check that you have at least Vite 3.2 installed

**Reply by ryinner** [reply]
That helped, but now i have another error: `No "useRouter" export is defined on the "/node_modules/vue-router/dist/vue-router.node.mjs"`

**Reply by sheremet-va** [reply]
It means that your code uses `useRouter` somewhere, but in your `vi.mock` function you did not specified it
